Allow users to set `email_address_visibility` during account creation.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit adds backend code to set `email_address_visibility` during account creation.
- Second commit adds UI to allow user to set `email_address_visibility` during account creation.

Fixes: <!-- Issue link, or clear description.-->#24310

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![email-visibility-nobody](https://user-images.githubusercontent.com/35494118/218992137-33931e0a-1274-40b2-948a-63ada79fe117.png)
![email-visibility-admins](https://user-images.githubusercontent.com/35494118/218992212-fd306dc1-d9bb-41cd-9e0d-ebeea514dd1e.png)
![email-visibility-others](https://user-images.githubusercontent.com/35494118/218992289-7b64eeb8-6921-4297-a366-c40f7be82a61.png)

![create-user-email-setting](https://user-images.githubusercontent.com/35494118/218992333-72ea386b-7a23-4d9e-acbe-824744a49078.gif)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Can we reduce the vertical spacing between the two lines of the notice?

timabbott commented on zerver/forms.py:
```diff
@@ -115,6 +115,7 @@ class RegistrationForm(forms.Form):
     realm_type = forms.IntegerField(required=False)
     is_demo_organization = forms.BooleanField(required=False)
     enable_marketing_emails = forms.BooleanField(required=False)
+    email_address_visibility = forms.IntegerField(required=False)
```
I think we need to somewhere verify that the value they selected is a valid `email_address_visibility` value; Django often likes to do that in the forms system, but I'm not sure specifically what the best way to do that is here.

sahil839 replied:
We can use `TypedChoiceFields` here. I have used that here for now. Other way can be to just add a function similar to `clean_full_name` which will just check whether the values are valid and then return the value as it is if they are valid and raise an error otherwise.

timabbott commented on zerver/views/registration.py:
```diff
@@ -523,6 +527,12 @@ def accounts_register(
             ),
+            "default_email_address_visibility": None
+            if realm is None
+            else realm_user_default.email_address_visibility,
```
This feels hard enough to read that I think we should compute a `default_email_address_visibility` above and just use its value here?

alya commented:
The dropdown styling looks a bit ugly to me. Is there something better we can reuse from elsewhere? I'd like to at least have more vertical space around the text.

![Screen Shot 2023-02-21 at 10 47 26 AM](https://user-images.githubusercontent.com/2090066/220432830-998f17d7-1b37-4258-887c-4d784b9fa563.png)

timabbott commented on zerver/views/registration.py:
```diff
@@ -327,6 +328,9 @@ def accounts_register(
         form["password"].field.required = False
 
+    if realm is not None:
+        realm_user_default = RealmUserDefault.objects.get(realm=realm)
```
Should this sit just before the `TemplateResponse` at the end? 

I think also we could make the type-checker a lot happier by having this block align with https://github.com/zulip/zulip/pull/24394/commits/7c278624fa249422d2c5b4923d8695d3550d8f75#r1113447984 to just have a single block of code that gives you an `default_email_address_visibility` value to put into the context out of it.

sahil839 replied:
Updated.

timabbott commented on zerver/models.py:
```diff
@@ -1611,13 +1611,31 @@ class UserBaseSettings(models.Model):
+            "value": EMAIL_ADDRESS_VISIBILITY_ADMINS,
+        },
+        EMAIL_ADDRESS_VISIBILITY_NOBODY: {
+            "name": "Nobody",
```
Do these strings need to be tagged for translation, perhaps with `gettext_lazy`? I think if your browser's default language is something other than English, and we do that, the options should be properly translated for users.

sahil839 replied:
It was tagged for translation in the template code similar to how organization-type choices are done. But I think using `gettext_lazy` here would be more cleaner, so updated it to use `gettext_lazy` instead.

timabbott commented on templates/zerver/register.html:
```diff
@@ -118,8 +118,26 @@ <h1>{{ _('Create your account') }}</h1>
+                            {% else %}
+                                {% trans %}Other users in this Zulip organization will be able to see this email address.
+                                {% endtrans %}
+                            {% endif %}
```
Reading these rules, maybe it's worth having an extra case for "Moderators of this ..." -- I can see that being meaningfully different for some users than "Other users" from a privacy standpoint.

(I don't think the guest/members/full members distinctions for this sort of setting are likely to be meaningful for a new user).

timabbott commented on templates/zerver/register.html:
```diff
@@ -118,8 +118,26 @@ <h1>{{ _('Create your account') }}</h1>
+                    <input type="hidden" name="email_address_visibility" id="email_address_visibility"/>
                     <label for="id_email" class="inline-block label-title">{{ _('Email') }}</label>
                     <div id="id_email">{{ email }}</div>
+                    {% if not creating_new_team %}
```
We should probably rename this to `creating_new_realm` to improve readability at some point.

sahil839 replied:
Will do this as a follow-up.

timabbott commented on static/js/portico/signup.js:
```diff
@@ -230,4 +233,36 @@ $(() => {
+        $("#email_address_visibility").val(selected_val);
+        Micromodal.close("change-email-address-visibility-modal");
+
+        let selected_option_text;
```
Let's add a comment here noting these strings are duplicated in the template, and should be kept in sync.

timabbott replied:
Or actually, could we make this text into a small Handlebars template taking an argument `option_code` that can be rendered here and included in the other place? Not sure if it makes things any simpler, since we'd need to pass the `settings_config` data in as a parameter. So up to you how to handle this.

sahil839 replied:
I am not sure whether we can or how to include a handlebars template inside an html file. So, added a comment for now.

timabbott replied:
Yeah that seems fine, I forgot this was still a Django template. I don't think we've really done much with handlebars outside the app.

alya commented:
If I change the default email visibility user setting while a user is on the "Create your account" page, clicking "Sign up" takes me to:

![Screen Shot 2023-02-21 at 10 53 14 AM](https://user-images.githubusercontent.com/2090066/220434063-a7afc7d5-7ea5-4161-b2d2-450e84407874.png)

alya commented:
That's all I've got. :)

alya commented:
Tested and it works fine if the default setting is changed while a user is going through the registration flow (once I checked by logging in as Desdemona in a separate browser).

sahil839 commented:
Updated the PR -
- Fixed the vertical spacing between lines.
- Also, fixed the dropdown design. It was due to recently added `modal_select` class not being present. Have kept the font-size for select elements same as rest of the modal which we do for other modals in the app too.
- Added code to show different notice for `Admins and moderators` case.
- Updated the code for tagging the options for translation.

![admins-moderators-default](https://user-images.githubusercontent.com/35494118/220710078-56a6de70-9961-45db-af14-c6561385f414.png)
![configure-email-visibility-modal](https://user-images.githubusercontent.com/35494118/220710109-d11c9735-68a2-4308-b125-bce1a71c8c86.png)

timabbott commented:
The left alignment of the new text looks off (it's a bit left of the options it's under)

![image](https://user-images.githubusercontent.com/2746074/220739312-5762c035-c7f4-43c4-971e-026a62472f9f.png)

The updated vertical spacing looks good though!

> We use Admins in the options in the app

Yeah, probably we should use "Admins" for consistency with the rest of the app. 

> So, what should we do when user is importing settings from another account while creating a new realm? Should we use the email_address_visibility of other account or just set the database default of Everyone?

I think we probably want to always be showing the user the value that we're going to use for this -- so probably the simplest story is that this setting is never copied from other accounts -- it always uses the value that's displayed in registration (or if a user is created in the API, the RealmUserDefault value -- but there will never be an existing account to copy from in that case anyway)

alya commented:
> Yeah, probably we should use "Admins" for consistency with the rest of the app.

Yeah, I think "Admins" in the dropdown, but "Administrators" in the other text.

alya commented:
> so probably the simplest story is that this setting is never copied from other accounts

I agree with that approach. We should tweak https://zulip.com/help/import-your-settings, so that it says something like:

- Privacy settings, except for email visibility, which you can configure when you join an organization.

alya commented:
Documentation PR: #24448

sahil839 commented:
Updated the PR. 
- Fixed the alignment of the text on the page. 
- Now we use `Admins` in options and `Administrators` in other text. Added one more field in `EMAIL_ADDRESS_VISIBILITY_TYPES_MAP` since we want different text in option and in the text above the dropdown in the modal.
![Screenshot from 2023-02-23 09-08-51](https://user-images.githubusercontent.com/35494118/220817099-36e7137e-76fc-4e9c-98eb-3c365de89ecf.png)

sahil839 commented:
> I think we probably want to always be showing the user the value that we're going to use for this -- so probably the simplest story is that this setting is never copied from other accounts -- it always uses the value that's displayed in registration (or if a user is created in the API, the RealmUserDefault value -- but there will never be an existing account to copy from in that case anyway)

Just to be clear, we do not show the `email_address_visibility` value in registration form when creating an organization and user can copy settings from existing account while creating new organization.

sahil839 commented:
Do want to point the help link with modal heading to `/help/configure-email-visibility`? We already have "after you join" pointing to that link.

alya commented:
> Do want to point the help link with modal heading to `/help/configure-email-visibility`? We already have "after you join" pointing to that link.

I think just one link is enough. The current way it works seems fine to me, or we could drop the link on "after you join" and put it in the modal heading instead.

sahil839 commented:
Ok. Keeping link on "after you join" seems better to me. Will keep it that way only.

timabbott commented:
This lgtm but I think we need a bit more copy-editing in the modal; awaiting updated text from @alya before integrating.

timabbott commented:
OK here's new language for the modal's text:

> ## Configure email address privacy

> Zulip lets you control which roles in the organization can view your email address. Do you want to change the privacy setting for your email from the default configuration for this organization?

> Who can access your email address

sahil839 commented:
Ok.

Just to confirm, we would still keep the text after the dropdown "You can also change this setting after you join", right?

timabbott commented:
This is great, merged, thanks @sahil839!
